### PR TITLE
Fix notifications sometimes not being sent

### PR DESCRIPTION
**Pull request by tomaka**
In the `poll` function of `handler.rs`, we first call `Sink::poll_flush` on all the outbound substreams, then only call `Sink::start_send`. This means that the actual flushing is only done the next time `poll` is called, which is in no way guaranteed to happen soon.

While this little fix might seem like a hack, and it would be preferable to change the order of operations in `poll`, doing so is a bit tricky to me because of having to handle state transitions when the substream errors. Since this code is undergoing another rewrite in #7374, I went for the easy solution.

### CODE REVIEW

File: client/network/src/protocol/generic_proto/handler.rs [hunk 1]
```diff
@@ -971,6 +971,7 @@ impl ProtocolsHandler for NotifsHandler {
 						if let Some(pos) = self.out_protocols.iter().position(|(n, _)| *n == protocol_name) {
 							if let Some(substream) = out_substreams[pos].as_mut() {
 								let _ = substream.start_send_unpin(message);
+								cx.waker().wake_by_ref();
```
Comment (mxinden) [inline client/network/src/protocol/generic_proto/handler.rs:974]: Would you mind documenting that this wake up is used to force another call to `NotifsHandler::poll` and thus to flush the corresponding substreams?

In addition I would suggest documenting that this is temporary and removed with https://github.com/paritytech/substrate/pull/7374.

**Review by mxinden — APPROVED** [review]
> While this little fix might seem like a hack, and it would be preferable to change the order of operations in poll, doing so is a bit tricky to me because of having to handle state transitions when the substream errors. Since this code is undergoing another rewrite in #7374, I went for the easy solution.

I am fine with this change as long as it is removed with https://github.com/paritytech/substrate/pull/7374/.

**Review by wheresaddie — APPROVED** [review]
I approve this based on it solving/removing #7374

**Comment by mxinden** [comment]
> I approve this based on it solving/removing #7374

@wheresaddie to make sure there is no confusion: This pull request does not make #7374 obsolete. Quite the opposite as #7374 would remove the _hack_ introduced in this pull request.

**Review by romanb — APPROVED** [review]

**Comment by tomaka** [comment]
bot merge

**Pull request merged**